python/decam_inject.py

```diff
@@ -0,0 +1,241 @@
+        rng = np.random.default_rng(int(date))
+        injextnames = rng.choice(injextnames, inject, replace=False)
+
+    # create files with injected sources in the decapsi directory
+    ## this might need to be more robust if we port to a different cluster/user
+    imfnI = injectRename(imfn)
+    ivarfnI = injectRename(ivarfn)
+    dqfnI = injectRename(dqfn)
+
+    # intialize the injected images
+    prihdr = fits.getheader(imfn, extname='PRIMARY')
+    if not resume or not os.path.exists(imfnI):
+        fits.writeto(imfnI, None, prihdr, overwrite=overwrite)
+
+    prihdr = fits.getheader(ivarfn, extname='PRIMARY')
+    if not resume or not os.path.exists(ivarfnI):
+        fits.writeto(ivarfnI, None, prihdr, overwrite=overwrite)
+
+    prihdr = fits.getheader(dqfn, extname='PRIMARY')
+    if not resume or not os.path.exists(dqfnI):
+        fits.writeto(dqfnI, None, prihdr, overwrite=overwrite)
+
+    import warnings
+    with warnings.catch_warnings(record=True) as wlist:
+        hdulist = fits.open(dqfnI)
+        injnamesdone = []
+        for hdu in hdulist:
+            if hdu.name == 'PRIMARY':
+                continue
+            injnamesdone.append(hdu.name)
+        hdulist.close()
+    # suppress endless nonstandard keyword warnings on read
+    for warning in wlist:
+        if 'following header keyword' in str(warning.message):
+            continue
+        else:
+            print(warning)
+
+    injextnamesI = [i+"I" for i in injextnames]
+    injextnamesI = [i for i in injextnamesI if i not in injnamescat]
+
+    injextnames = [i for i in injextnames if i not in injnamesdone]
+
+    rng = np.random.default_rng(int(date))
+    for key in injextnames:
+        scatter_stars(outfn, imfn, ivarfn, dqfn, key, filt, pixsz, wcutoff, verbose, rng, injectfrac=injectfrac)
+
+    return imfnI, ivarfnI, dqfnI, injextnamesI
+
+#seed on date here too
+def scatter_stars(outfn, imfn, ivarfn, dqfn, key, filt, pixsz, wcutoff, verbose, rng, injectfrac=0.1):
+
+    ## imports
+    hdr = fits.getheader(outfn,key+"_HDR")
+    gain = hdr['GAINCRWD']
+
+    f = fits.open(outfn)
+    table = f[key+"_CAT"].data
+    flux_stars = table["flux"]
+    flags_stars = table["flags"]
+    f.close()
+
+    psfmodel = load_psfmodel(outfn, key, filt[0], pixsz=pixsz)
+
+    import warnings
+    with warnings.catch_warnings(record=True) as wlist:
+        im = fits.getdata(imfn, extname=key).copy()
+        wt = fits.getdata(ivarfn, extname=key).copy()
+        dq = fits.getdata(dqfn, extname=key).copy()
+    # suppress endless nonstandard keyword warnings on read
+    for warning in wlist:
+        if 'following header keyword' in str(warning.message):
+            continue
+        else:
+            print(warning)
+
+    nx, ny = im.shape
+
+    # this requres stars to be "good" and in a reasonable flux range (0 flux to 17th mag)
+    maskf = ((flags_stars==1) | (flags_stars==2097153)) & (flux_stars>0) & (flux_stars<158489.3192461114);
```

Comment:
Is this 17th mag assuming some nominal zero point for a particular band?  This number is oddly specific.

Reply:
LOL, that is exactly what it is. It is 17th mag in g-band using the means zero point from DECaPS 1. I don't think the actual number matters, that is just how I picked it.

Reply:
Let's clarify both these numbers after your run.  i.e., 150000 to me better expresses the concept of "not too bright" than 158489; the corresponding comment might be that 17th mag in g roughly corresponds to 160k counts.

I'd usually write your flag cut as  something like (flags & ~decam_proc.extrabits['diffuse']) == 2**0, but that's just stylistic.  It would be good form to refer to the diffuse bit by name rather than by that decimal.

Reply:
Sure. Good code comments all.